### ISSUE TITLE
Docker Compose resource

### ISSUE THREAD

**Issue by oppegard** [concourse/concourse#324]
External services (MySQL, redis, memcached) and the need for an "uber container" to run our test suites are a pain point for the Pivotal Tracker team. For example, our Dockerfile installs the same version of MySQL 5.6.x that we run on pivotaltracker.com to run tests. But, we also have a [PCF tile](https://network.pivotal.io/products/pivotal-tracker-cf) that runs against [p-mysql](https://network.pivotal.io/products/p-mysql). There seem to be two options for testing both DBs:

1) Have a second docker image that duplicates the uber image, substituting MariaDB 10.0.x for MySQL 5.6.x.

2) Jump through hoops of installing both MySQL and MariaDB in the same docker image (using something like http://mysqlsandbox.net/)

We've gone with option 1 for the time being. While an external service such as RDS is an option, I'd rather not be in the business of managing pools of instances and incur the extra network latency.

----

In an ideal, future-Concourse world, I could mix and match my external services with Docker Compose. As an example, see how Logplex runs their suite on TravisCI:

* https://github.com/heroku/logplex/blob/master/.travis.yml#L13-L15

If they want to run against redis 3.x, it's as simple as changing a line: https://github.com/heroku/logplex/blob/master/docker-compose.yml#L5

An added benefit of Docker Compose is how quickly you get an environment running locally on your workstation. I don't have any of the Logplex dependencies on my MacBook (erlang, redis, etc.). But, my onboarding experience was:

1. `git clone git@github.com:heroku/logplex.git`
2. `cd logplex`
3.  `docker-compose build && docker-compose run test`

In the span of 5 minutes I had an erlang test suite running against redis 2.8.

**Comment by bechtoldt** [comment]
I need i.e. to bootstrap some containers to run integration tests. Is this somehow possible yet?

**Comment by ahelal** [comment]
Or some thinking like [Travis](https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix) matrix or test kitchen suites. So you can run tests with different combination of things i.e. version, os's, ...

**Comment by rajatvig** [comment]
Is this support likely to land anytime soon?

**Comment by oppegard** [comment]
It'd be nice to hear from @vito if something like this fits into the conceptual model of Concourse. I haven't thought through what supporting Docker Compose means, only that I want it. For example, I'm not sure if this makes sense as a Resource. To be more clear about what I'd like:

- A task.yml can specify "service resources" as additional containers that are started along with the main build container. All the containers in a build are networked to each other.
- Running a one-off build via `fly execute` still works.

It seems like a `docker-compose.yml` would be a good way to describe the desired containers and their networking. It lets teams use Concourse for CI and Docker Compose for local development.

**Comment by vito** [comment]
When I have more time I want to play with allowing resources to expose network addresses to builds. That way you could model throwaway databases and services as resources that go away when the build is done. This could be done either with a single Docker Compose resource or with individual postgres/mysql/etc. resources.

**Comment by Tankanow** [comment]
+1 
I vote for as-native-as-possible `docker-compose` support. One simple sounding, if not implementable, solution is to give tasks access to the worker OS.

**Comment by royseto** [comment]
Hi, I learned about Concourse this morning and like almost everything I've read about it. Currently we use Jenkins and have a build pipeline that first builds a Docker image, then runs a bunch of test suites against that image in downstream jobs. 

However, I think not being able to link dependent containers into a single Concourse build job will block us from migrating to Concourse. Our app is a Python/Flask app with dependencies on postgres and redis. My build script needs to look something like this:

```
docker-compose up
docker-compose run runtests /tests.sh
```

where my docker-compose.yml looks something like this:

```
runtests:
  image: testbuild
  command: /bin/true
  volumes:
   - ./local.py:/app/config/local.py
   - ./runtests.sh:/runtests.sh
  links:
   - db
   - redis
db:
  image: my.docker.registry/postgres-dbimage
redis:
  image: redis
```

I want to containerize postgres and redis so that we can parallelize our tests and avoid having them step on each other.

Please let me know if there's a way to accomplish what I am trying to do in Concourse today. Unless I'm missing something, I think I need to make a shell script that does the above and keep calling it from Jenkins for now. Thanks!

**Comment by vito** [comment]
With 1.3 it'll be pretty trivial to run Docker within your task:

```sh
docker daemon &
docker-compose up
docker-compose runtests /tests.sh
```

This can already be done today, with some workarounds that won't be necessary come 1.3 (e.g. mounting a tmpfs for the Docker graph). You'd also need to run the task with `privileged: true`, which may also be fixable now that cgroup namespacing is supported in Linux 4.6, though that will have to wait for Garden-runC to support it (/cc @julz).

**Comment by royseto** [comment]
Thanks @vito. I'll try it out after 1.3 comes out. Do you have a sense for what the performance overhead would be of running Docker containers inside Garden instead of running just Docker or Garden inside the host machine?

**Comment by julz** [comment]
There shouldn't really be much of an impact: containers are just processes (there's no virtualisation going on), the overhead is very very small. The only real impact I can think of is you don't get the shared image cache from the host.

**Comment by royseto** [comment]
OK, thanks for clarifying that @julz.

**Comment by seanknox** [comment]
> With 1.3 it'll be pretty trivial to run Docker within your task:
> 
> docker daemon &
> docker-compose up
> docker-compose runtests /tests.sh
> 
> This can already be done today, with some workarounds that won't be necessary come 1.3 (e.g. mounting a tmpfs for the Docker graph). You'd also need to run the task with privileged: true, which may also be fixable now that cgroup namespacing is supported in Linux 4.6, though that will have to wait for Garden-runC to support it (/cc @julz).

@julz @vito, short of docker-compose or concourse v1.3 availability, how does one run tasks that need to run multiple, linked Docker containers? E.g. a Rails app container that also needs additional service containers running like Redis and Postgres.

**Comment by drnic** [comment]
You could run a pool of pre-existing docker VMs and target docker-compose
against them. This is a similar model to deploying/testing bosh deployments
against an external bosh-lite.

**Comment by Tankanow** [comment]
@drnic, can you explain this a bit more? How do I integrate this pool of VMs into concourse?

**Comment by vito** [comment]
@seanknox That sounds like something you'd just use docker-compose for. I'm not as familiar with it as I should be though - I personally prefer to have my test suites self-contained and spin up their own Postgres server (e.g. [here](https://github.com/concourse/atc/blob/master/db/db_suite_test.go#L23)) - but that's purely a stylistic thing.

**Comment by simonvanderveldt** [comment]
> With 1.3 it'll be pretty trivial to run Docker within your task:
> 
> docker daemon &
> docker-compose up
> docker-compose runtests /tests.sh
> This can already be done today, with some workarounds that won't be necessary come 1.3 (e.g. mounting a tmpfs for the Docker graph). You'd also need to run the task with privileged: true, which may also be fixable now that cgroup namespacing is supported in Linux 4.6, though that will have to wait for Garden-runC to support it (/cc @julz).

@vito This means you'd need an image that contains both the docker binary as well as the docker-compose binary to run these commands in, correct?
And because you're running docker in a container there won't be any image caching between runs, correct?

Would it be possible to pass in the docker binary as well as the docker socket of the host the worker is on? That way you can make use of the docker daemon on the host and it's caching.
Or alternatively I guess garden(-runc) would need to support Docker Compose somehow?

**Comment by vito** [comment]
@simonvanderveldt There is no Docker binary or socket on the host - they're just running a Garden backend (probably Guardian). Concourse runs at an abstraction layer above Docker, so providing any sort of magic there doesn't really make sense.

The one thing missing post-1.3 is that Docker requires you to set up cgroups yourself. I forgot how annoying that is. I wish they did what Guardian does and auto-configure it, but what can ya do.

So, the full set of instructions is:

1. Use or a build an image with `docker` in it, e.g. `docker:dind`.
1. Run the following at the start of your task: https://github.com/concourse/docker-image-resource/blob/master/assets/common.sh#L1-L40
1. Spin up Docker with `docker daemon &`.

Then you can run `docker-compose` and friends as normal.

The downside of this is that you'll be fetching the images every time. https://github.com/concourse/concourse/issues/230 will address that.

In the long run, https://github.com/concourse/concourse/issues/324#issuecomment-212505738 is the direction I want to go.

**Comment by mumoshu** [comment]
> an image that contains both the docker binary as well as the docker-compose binary to run these commands in

@simonvanderveldt FYI, I've created a `dcind` docker image for it https://github.com/mumoshu/dcind

It's working for me and my colleagues.

One caveat is that, to prevent `docker-compose` from fetching all the dependent docker images from scratch on each task run, you have to provide appropriate `docker-image-resource` inputs to the task, e.g. if your compose.yml references the `redis` image, provide that image via an input and `docker load` it like https://github.com/mumoshu/concourse-aws/blob/master/ci/tasks/compose.yml#L25

As @vito mentioned, once https://github.com/concourse/concourse/issues/230 lands, this `docker load` will be unnecessary.

**Comment by meAmidos** [comment]
@mumoshu Thank you for the ```dcind```! It turned out to be a very useful solution, and I used it as a reference for a slightly simplified image: https://github.com/meAmidos/dcind

**Comment by martinsbalodis** [comment]
I am having the same issue as everyone else in here. I would suggest a solution that is similar to travis.ci approach. Basically the required services for the test could be spawned as a docker containers that are accessible via network via dns names. For example a build.yml might look like this:

```yml
platform: linux

image_resource:
  type: docker-image
  source: {repository: busybox}

services:
  - name: sql (acessible via dns by resolving sql)
    type: docker-image
    source: {repository: mysql}
  - name: rabbitmq (acessible via dns by resolving rabbitmq)
    type: docker-image
    source: {repository: rabbitmq}

inputs:
- name: flight-school

run:
  path: ./flight-school/ci/test.sh
```

**Comment by MaikuMori** [comment]
I'd like to second the service (sidecar) idea already mentioned by @martinsbalodis . I believe it's the most universal solution.

**Comment by poida** [comment]
As an alternative to putting the services in the task you could define it in the pipeline and have it pass the `network` to the task (I guess like what @vito was talking about above by capturing it in a resource).

Pipeline:
```yaml
jobs:
  - name: integ-test
    serial: true
    plan:
      - get: source
        trigger: true
      - put: compose-integ-test
        network: true
        params:
          docker-compose: source/integ-tests/compose
      - task: integ-test
        network: compose-integ-test
        ensure:
          put: compose-integ-test
          params:
            stop: true
```

Task: integ-test
```yaml
platform: linux

image_resource:
  type: docker-image
  source: {repository: busybox}

inputs:
- name: source

run:
  path: source/ci/integ-test.sh
```

**Comment by hryamzik** [comment]
Let me add my 5 cents: services are a good pattern for some integration tests until you have to test integration between two products both developed by you team. They might not mach the services idea but could require different linked docker containers or hardware workers to run. Simple example is a mobile app tested against server side code.

**Comment by devdavidkarlsson** [comment]
You guys who run something like: 

docker daemon &
docker-compose up
docker-compose runtests /tests.sh

How do you make the task fail when the tests are failing?

**Comment by smazurov** [comment]
typically `tests.sh` would exit with a non-zero status.

**Comment by crstamps2** [comment]
I wonder if now that Docker Compose has links, we could leverage something like this to do things such as running a rails app in a docker container but linked to a separate mysql docker container. I visualize it as both docker containers being a resource for a job/task.

**Comment by martinsbalodis** [comment]
While this is being resolved I decided to create a workaround - a single container for unit testing which would contain all dependency services. Here is an example Dockerfile with some concourse relevant configuration samples https://github.com/martinsbalodis/docker-concourse-php-unittests .

**Comment by artem-zinnatullin** [comment]
>The only real impact I can think of is you don't get the shared image cache from the host.

This is an issue for us since our images are about 5-7 GB… Any possible workarounds except docker registry instance in the ~same network?

**Comment by pecigonzalo** [comment]
FYI, was having issues with all this, i created https://hub.docker.com/r/pecigonzalo/docker-ecr/ which is similar to the other dind images but based on regular install and debian, still sort of WIP for testing only, but let me know.
Includes
`docker`
`docker-compose`
`amazon-ecr auth helper` 

We need to ensure to stop docker at the end, its ugly but it gets the job done

```
- task: sbt-build-image
      privileged: true
      config:
        platform: linux
        image_resource:
          type: docker-image
          source:
            repository: pecigonzalo/docker-ecr
            tag: latest
        run:
          path: sh
          args:
          - -exc
          - |
            service docker start
            service docker status
            sleep 5
            docker load -i ./myimage/image
            docker tag "$(cat ./myimage/image-id)" "$(cat ./myimage/repository):$(cat ./myimage/tag)"
            docker images
            cd gitrepo
            docker-compose up -d
            docker images
            service docker stop
          dir: ""
        inputs:
        - name: twoflower-jar
          path: ""
        - name: sbt-image
          path: ""
```

**Comment by thanizebra** [comment]
@pecigonzalo 
here is my log.

+ service docker start
mount: cgroup is already mounted or /sys/fs/cgroup/cpu busy
       cgroup is already mounted on /sys/fs/cgroup
       cgroup is already mounted on /sys/fs/cgroup/cpuset
mount: cgroup is already mounted or /sys/fs/cgroup/cpuacct busy
       cgroup is already mounted on /sys/fs/cgroup
       cgroup is already mounted on /sys/fs/cgroup/cpuset
mount: cgroup is already mounted or /sys/fs/cgroup/net_cls busy
       cgroup is already mounted on /sys/fs/cgroup
       cgroup is already mounted on /sys/fs/cgroup/cpuset
       cgroup is already mounted on /sys/fs/cgroup/blkio
       cgroup is already mounted on /sys/fs/cgroup/memory
       cgroup is already mounted on /sys/fs/cgroup/devices
       cgroup is already mounted on /sys/fs/cgroup/freezer
mount: cgroup is already mounted or /sys/fs/cgroup/net_prio busy
       cgroup is already mounted on /sys/fs/cgroup
       cgroup is already mounted on /sys/fs/cgroup/cpuset
       cgroup is already mounted on /sys/fs/cgroup/blkio
       cgroup is already mounted on /sys/fs/cgroup/memory
       cgroup is already mounted on /sys/fs/cgroup/devices
       cgroup is already mounted on /sys/fs/cgroup/freezer
       cgroup is already mounted on /sys/fs/cgroup/perf_event
Starting Docker: docker.
+ service docker status
Docker is running.
+ sleep 5
+ docker load -i ./nginx/image
d17d48b2382a: Loading layer 128.9 MB/128.9 MB
cf0d2468e726: Loading layer  60.9 MB/60.9 MB
e3a86ffb1c45: Loading layer 3.584 kB/3.584 kB
Loaded image ID: sha256:6b914bbcb89e49851990e064568dceee4d53a462f316ec36207599c12ae9ba65
+ cat ./nginx/image-id
+ cat ./nginx/repository
+ cat ./nginx/tag
+ docker tag sha256:6b914bbcb89e49851990e064568dceee4d53a462f316ec36207599c12ae9ba65 nginx:latest
+ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
nginx               latest              6b914bbcb89e        10 days ago         182 MB
+ cd gitrepo
sh: 7: cd: can't cd to gitrepo


if running command is failed then the task should stop and show failure.
bu here it is just hanging for long time

**Comment by pecigonzalo** [comment]
@thanizebra yeah the stop issue is something i have to yet figure out, i have the same problem i realized.

Regarding the mounts, what is your setup? where are you running the containers?

**Comment by thanizebra** [comment]
@pecigonzalo : sorry for my delayed response (was on holiday)
mount was working fine 
only problem is stop issue.

if there is no error occurred then everything working fine
i have checked a demo rails app with postgresql and angularjs.
i can able to build successfully if error not occured.

here is my setup
---
platform: linux
image_resource:
  type: docker-image
  source: {repository: pecigonzalo/docker-ecr, tag: latest}

inputs:
- name: thani_api_repo
- name: pipelines_repo
- name: thani_api_dev_image
params:
  THANI_API_REPO:
  PIPELINES_REPO:
  THANI_API_DEV_IMAGE:

run:
  path: sh
  args:
  - -exc
  - |
    service docker start
    service docker status
    sleep 5
    docker load -i ./thani_api_dev_image/image
    docker tag "$(cat ./thani_api_dev_image/image-id)" "$(cat ./thani_api_dev_image/repository):$(cat ./thani_api_dev_image/tag)"
    #export IMG="$(cat ./thani_api_dev_image/repository):$(cat ./thani_api_dev_image/tag)"
    docker images
    cd thani_api_repo
    docker-compose up -d
    sleep 5
    docker images
    docker-compose down
    service docker stop
    rc=$?
    cd ..
    echo "exit code = $rc "
    #kill %1
    exit $rc
  dir: ""

**Comment by pecigonzalo** [comment]
I updated the image some hours ago, try running as 
```
      run:
        path: /bin/sh
        args:
          - -exc
          - |
              . /root/common.sh
              start_docker
              docker ps
              docker-compose version
```

**Comment by devdavidkarlsson** [comment]
http://www.davidkarlsson.no/2016/12/26/integration-testing-in-concourse-ci/ my Docker in docker setup

**Comment by thaniyarasu** [comment]
@pecigonzalo : Here is my log output for a simple service

+ . /root/common.sh
+ start_docker
+ mkdir -p /var/log
+ mkdir -p /var/run
+ sanitize_cgroups
+ mkdir -p /sys/fs/cgroup
+ mountpoint -q /sys/fs/cgroup
+ mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup /sys/fs/cgroup
+ mount -o remount,rw /sys/fs/cgroup
+ sed -e 1d /proc/cgroups
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ grep \<cpuset\>
+ cut -d: -f2
+ grouping=cpuset
+ [ -z cpuset ]
+ mountpoint=/sys/fs/cgroup/cpuset
+ mkdir -p /sys/fs/cgroup/cpuset
+ mountpoint -q /sys/fs/cgroup/cpuset
+ mount -n -t cgroup -o cpuset cgroup /sys/fs/cgroup/cpuset
+ [ cpuset != cpuset ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ grep \<cpu\>
+ cut -d: -f2
+ grouping=cpu,cpuacct
+ [ -z cpu,cpuacct ]
+ mountpoint=/sys/fs/cgroup/cpu,cpuacct
+ mkdir -p /sys/fs/cgroup/cpu,cpuacct
+ mountpoint -q /sys/fs/cgroup/cpu,cpuacct
+ mount -n -t cgroup -o cpu,cpuacct cgroup /sys/fs/cgroup/cpu,cpuacct
+ [ cpu,cpuacct != cpu ]
+ [ -L /sys/fs/cgroup/cpu ]
+ ln -s /sys/fs/cgroup/cpu,cpuacct /sys/fs/cgroup/cpu
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ + cut -d: -f2
grep \<cpuacct\>
+ grouping=cpu,cpuacct
+ [ -z cpu,cpuacct ]
+ mountpoint=/sys/fs/cgroup/cpu,cpuacct
+ mkdir -p /sys/fs/cgroup/cpu,cpuacct
+ mountpoint -q /sys/fs/cgroup/cpu,cpuacct
+ umount /sys/fs/cgroup/cpu,cpuacct
+ mount -n -t cgroup -o cpu,cpuacct cgroup /sys/fs/cgroup/cpu,cpuacct
+ [ cpu,cpuacct != cpuacct ]
+ [ -L /sys/fs/cgroup/cpuacct ]
+ ln -s /sys/fs/cgroup/cpu,cpuacct /sys/fs/cgroup/cpuacct
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<blkio\>
+ grouping=blkio
+ [ -z blkio ]
+ mountpoint=/sys/fs/cgroup/blkio
+ mkdir -p /sys/fs/cgroup/blkio
+ mountpoint -q /sys/fs/cgroup/blkio
+ mount -n -t cgroup -o blkio cgroup /sys/fs/cgroup/blkio
+ [ blkio != blkio ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<memory\>
+ grouping=memory
+ [ -z memory ]
+ mountpoint=/sys/fs/cgroup/memory
+ mkdir -p /sys/fs/cgroup/memory
+ mountpoint -q /sys/fs/cgroup/memory
+ mount -n -t cgroup -o memory cgroup /sys/fs/cgroup/memory
+ [ memory != memory ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat+ cut -d: -f2
+ grep \<devices\>
 /proc/self/cgroup
+ grouping=devices
+ [ -z devices ]
+ mountpoint=/sys/fs/cgroup/devices
+ mkdir -p /sys/fs/cgroup/devices
+ mountpoint -q /sys/fs/cgroup/devices
+ mount -n -t cgroup -o devices cgroup /sys/fs/cgroup/devices
+ [ devices != devices ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ + cutgrep -d: -f2
 \<freezer\>
+ grouping=freezer
+ [ -z freezer ]
+ mountpoint=/sys/fs/cgroup/freezer
+ mkdir -p /sys/fs/cgroup/freezer
+ mountpoint -q /sys/fs/cgroup/freezer
+ mount -n -t cgroup -o freezer cgroup /sys/fs/cgroup/freezer
+ [ freezer != freezer ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<net_cls\>
+ grouping=net_cls,net_prio
+ [ -z net_cls,net_prio ]
+ mountpoint=/sys/fs/cgroup/net_cls,net_prio
+ mkdir -p /sys/fs/cgroup/net_cls,net_prio
+ mountpoint -q /sys/fs/cgroup/net_cls,net_prio
+ mount -n -t cgroup -o net_cls,net_prio cgroup /sys/fs/cgroup/net_cls,net_prio
+ [ net_cls,net_prio != net_cls ]
+ [ -L /sys/fs/cgroup/net_cls ]
+ ln -s /sys/fs/cgroup/net_cls,net_prio /sys/fs/cgroup/net_cls
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<perf_event\>
+ grouping=perf_event
+ [ -z perf_event ]
+ mountpoint=/sys/fs/cgroup/perf_event
+ mkdir -p /sys/fs/cgroup/perf_event
+ mountpoint -q /sys/fs/cgroup/perf_event
+ mount -n -t cgroup -o perf_event cgroup /sys/fs/cgroup/perf_event
+ [ perf_event != perf_event ]
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<net_prio\>
+ grouping=net_cls,net_prio
+ [ -z net_cls,net_prio ]
+ mountpoint=/sys/fs/cgroup/net_cls,net_prio
+ mkdir -p /sys/fs/cgroup/net_cls,net_prio
+ mountpoint -q /sys/fs/cgroup/net_cls,net_prio
+ umount /sys/fs/cgroup/net_cls,net_prio
+ mount -n -t cgroup -o net_cls,net_prio cgroup /sys/fs/cgroup/net_cls,net_prio
+ [ net_cls,net_prio != net_prio ]
+ [ -L /sys/fs/cgroup/net_prio ]
+ ln -s /sys/fs/cgroup/net_cls,net_prio /sys/fs/cgroup/net_prio
+ read sys hierarchy num enabled
+ [ 1 != 1 ]
+ cat /proc/self/cgroup
+ cut -d: -f2
+ grep \<pids\>
+ grouping=pids
+ [ -z pids ]
+ mountpoint=/sys/fs/cgroup/pids
+ mkdir -p /sys/fs/cgroup/pids
+ mountpoint -q /sys/fs/cgroup/pids
+ mount -n -t cgroup -o pids cgroup /sys/fs/cgroup/pids
+ [ pids != pids ]
+ read sys hierarchy num enabled
+ grep /proc/sys\s\+\w\+\s\+ro, /proc/mounts
+ ip route get 8.8.8.8
+ awk { print $5 }
+ cat /sys/class/net/wb116shba3me-1/mtu
+ local mtu=1500
+ local server_args=--mtu 1500
+ local registry=
+ [ -n  ]
+ echo 109
+ trap stop_docker EXIT
+ sleep 1
+ docker daemon --mtu 1500
+ docker info
+ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
+ docker-compose version
docker-compose version 1.11.2, build dfed245
docker-py version: 2.1.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
+ stop_docker
+ cat /tmp/docker.pid
+ local pid=109
+ [ -z 109 ]
+ kill -TERM 109
+ wait 109

**Comment by thaniyarasu** [comment]
@pecigonzalo : why don't you prefer to use the existing image ?? i mean why are extracting rootfs.tar instead of using image. which you obtain by using  
---
get: nginx
params: {save:true, roofs: true}
---
here is list of files under that nginx image
---
+ ls -lart nginx
total 369148
-rw------- 1 root root 189815808 Mar 16 20:41 image
-rw-r--r-- 1 root root        72 Mar 16 20:41 image-id
-rw-r--r-- 1 root root      3203 Mar 16 20:41 docker_inspect.json
-rw-r--r-- 1 root root       138 Mar 16 20:41 metadata.json
drwxr-xr-x 1 root root       152 Mar 16 20:41 rootfs
-rw-r--r-- 1 root root         7 Mar 16 20:41 tag
-rw-r--r-- 1 root root 188163072 Mar 16 20:41 rootfs.tar
-rw-r--r-- 1 root root        57 Mar 16 20:41 repository
-rw-r--r-- 1 root root        72 Mar 16 20:41 digest
drwxr-xr-x 1 root root       160 Mar 16 20:41 .
drwxr-xr-x 1 root root        62 Mar 16 20:45 ..

**Comment by pecigonzalo** [comment]
@thaniyarasu I dont understand your question...

**Comment by thaniyarasu** [comment]
@pecigonzalo 
get: docker-image-resource-nginx
params: {save: true, rootfs: save}

that will store both "image" and "rootfs". 
can't you utilize "image" without "untar rootfs" ?

**Comment by pecigonzalo** [comment]
When/Where am i doing that?

**Comment by thanizebra** [comment]
@pecigonzalo : at here
            docker load -i ./myimage/image
            docker tag "$(cat ./myimage/image-id)" "$(cat ./myimage/repository):$(cat ./myimage/tag)"

see here you are loading the image again and taging it into task container.
can't we use existing image without manually adding into task container.
because i have to add more than 4 image for integration test. just asking if there is any possibility ??

**Comment by pecigonzalo** [comment]
@thanizebra Ok, it was a bit confusing because he said:
> why don't you prefer to use the existing image ?? i mean why are extracting rootfs.tar instead of using 'image'.
Which seems to refer to something else, and we are actually loading the image on the snippet you pasted.

The problem is we need to load the image into the docker image "repository". In your scenario, it might be better to have an external "proxy" docker repository to cache the downloads etc.
EG: https://blog.docker.com/2015/10/registry-proxy-cache-docker-open-source/


As far as i understand, there is no other way to "cache images" unless we change how concourse is working to store the images differently.

**Comment by thanizebra** [comment]
@devdavidkarlsson 
is this correct url  ?
http://www.davidkarlsson.no/2016/12/26/integration-testing-in-concourse-ci/
not working .

**Comment by devdavidkarlsson** [comment]
Some temporary down time, @thanizebra try now...

**Comment by dsyer** [comment]
I tried following the pattern in http://www.davidkarlsson.no/2016/12/26/integration-testing-in-concourse-ci/. It looks like a decent idea but isn't working for me yet. I can run everything I need on my local docker daemon, but the non-docker environment in concourse seems to behave differently, and I can't see how to fix it (or even if it's possible to fix it).

When I try to start `dockerd` in the concourse worker it fails like this:

```
# dockerd --storage-driver=vfs --config-file=/etc/docker/daemon.json -p /var/run/docker-bootstrap.pid
WARN[0000] could not change group /var/run/docker.sock to docker: group docker not found 
DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock) 
INFO[0000] libcontainerd: new containerd process, pid: 148 
DEBU[0000] containerd: grpc api on /var/run/docker/libcontainerd/docker-containerd.sock 
DEBU[0000] containerd: read past events                  count=0
DEBU[0000] containerd: supervisor running                cpus=4 memory=7983 runtime=docker-runc runtimeArgs=[] stateDir="/var/run/docker/libcontainerd/containerd"
DEBU[0000] libcontainerd: containerd health check returned error: rpc error: code = 14 desc = grpc: the connection is unavailable 
WARN[0001] failed to rename /var/lib/docker/tmp for background deletion: %!s(<nil>). Deleting synchronously 
DEBU[0001] Using default logging driver json-file       
DEBU[0001] Golang's threads limit set to 57240          
DEBU[0001] [graphdriver] trying provided driver: vfs    
DEBU[0001] Using graph driver vfs                       
DEBU[0001] Max Concurrent Downloads: 3                  
DEBU[0001] Max Concurrent Uploads: 5                    
INFO[0001] Graph migration to content-addressability took 0.00 seconds 
WARN[0001] Your kernel does not support cgroup memory limit 
WARN[0001] Unable to find cpu cgroup in mounts          
WARN[0001] Unable to find blkio cgroup in mounts        
WARN[0001] Unable to find cpuset cgroup in mounts       
WARN[0001] mountpoint for pids not found                
DEBU[0001] Cleaning up old mountid : start.             
Error starting daemon: Devices cgroup isn't mounted
```

I tried various different options for `--storage-driver`, having noted that the blog link above used `bitrfs`, but none of them worked, and all except `vfs` resulted in errors in the logs. It doesn't make any difference if I add `privileged: true` to the task.

**Comment by devdavidkarlsson** [comment]
Sorry it is not working for you @dsyer. What docker image are you using where I specify:` image: docker-registry-dind`

**Comment by dsyer** [comment]
I was using `docker:dind`

**Comment by dajulia3** [comment]
Hey @vito I just want to resurface this issue as a number of teams and clients here at Pivotal Chicago have been clamoring for this. I've also personally felt pain around this, and it is seen as a primary disadvantage compared to Travis or Gitlab when our clients are discussing which tool to use. If there's anything we can do to help the concourse team understand the use cases, let me know!

**Comment by pshomov** [comment]
Hello,

I understand mapping `/var/run/docker.sock` might not be the cleanest solution but I have been using it for the longest times and solves all my problems - simplicity and performance (although, I am guessing, at some security risk). 
Is there a way to tell the task to map a volume to the docker container I am using so I can do the equivalent of `-v /var/run/docker.sock:/var/run/docker.sock`?

Best regards,

Petar

**Comment by josebarn** [comment]
@dajulia3 I updated an example from @meAmidos and is located here: https://github.com/josebarn/dcind

The example should work out of the box. 

It is really clean - and allows the images which used by the docker-compose.yml file to be incoming resources (per the paradigm of concourse).

So it should be easy to use AWS images for example w/o having to jump through hoops in the compose file.

I have to really stretch the legs on this approach before I declare this completely solid - but check it our for yourself. :)

**Comment by EugenMayer** [comment]
Drone.io implements this and since its open-sourced, maybe its worth trying to fish for and implementation there. They do use the services-syntax exactly the same way as stated initially here , and they are written in go - so at least we could reuse the BL of it, fitting it into the cci framework.

Usually i would tend to say that running the stack for smoke-tests, integration tests and also acceptance tests is off the grid for a CI too. It should be part of a deployment plan, be it the whole stack or just a microservice, deploying a docker-compose with all you need and then running a test against it. You could even ship the images using save/load without putting it into your private repo yet ( even though, that is unpractical due to the time save/load takes)

Sure that is a cool feature to have in a CI tool, since you can test the images right were you created them, on the other side, CI agents tend to have not the right specs to run thus applications and produce issues which staging/production nodes would not. Its probably wishing for something which will never be quiet perfect.

**Comment by pecigonzalo** [comment]
Based on @josebarn example and some other examples, here is one based on `docker:dind` and also with autoloading of all images placed under `images/`.
This simplifies the loading process by just putting all images in a specific path.

https://github.com/pecigonzalo/docker-concourse-dind

**Comment by hukl** [comment]
I was just looking for that ability. Right now we're using CircleCI for a project, which offers this feature as well. As this issue is still unresolved with Concourse, we'll stick with Circle for now :/

**Comment by willejs** [comment]
@vito is this on the roadmap? Doing dind with docker compose isnt very nice at all. I am having issues on larger projects using this method.

**Comment by vito** [comment]
@willejs Sorry, this is currently not on the roadmap. We're focusing right now on spaces (#1707) and RBAC (#1317) and don't have the bandwidth to start this in parallel. I do want to see discussion/planning on this revived. Maybe people could submit RFCs (https://github.com/concourse/rfcs) with their own mock-ups and we can see where things go from there?

**Comment by jchesterpivotal** [comment]
As an aside, this feature might become simpler to implement as/when Kubernetes takes over as the container management layer: https://github.com/kubernetes/kompose

**Comment by vito** [comment]
I don't think task caches work very well for that as it won't preserve the
mount config across task runs.

**Comment by pswenson** [comment]
in my old CI systems (jenkins and drone) my workflow was the following:
1. checkout out code
2. run unit tests
3. build docker image
4. using docker-compose run a smoke test against the docker image that was just built (hit health check plus send a few messages to http endpoint)
5. publish image

I'm finding that with concourse it appears the smoke test is not possible unless I use DinD, which has significant downsides.  It's very unfortunate that I have to publish a possibly bad image to the registry (just because unit tests pass doesn't meant that starting a container will work).

Is there an alternative that I don't know about?

thanks

**Comment by AnianZ** [comment]
@pswenson I've had good experiences with [a version of the DinD image](https://github.com/cioplenu/dcind) 

My pipeline pulls in the git resouce, does a set of unit tests and linting first, if those passes it triggers  a build step that build a docker container and pushes it with a special tag to the registry. If the build is successful that image triggers two more jobs that run some smoke and and integration tests against multiple containers spun up with docker-compose including a empty postgres db which just needed some small tweaks to run faster in that setup. A essential pice of this setup is [wait-for-it](https://github.com/vishnubob/wait-for-it) 

My team is not especially big but I've run more over 1000 of each of these jobs without major problems. If the jobs fail for reasons other then a real problem with the build it's the concourse worker running out of IOPS if too many jobs run on it in parallel but even that is very rare with the database in docker-compose configured so it does not fsync.

It's not optimal but very workable for my use case right now. But of course I would still appreciate official support for something like this.

**Comment by trulede** [comment]
> rs the smoke test is not possible unless I use DinD, which has significant downsides. It's very unfortunate that I have to publish a possibly bad image to the registry.
> 
> just because unit tests pass doesn't meant that starting a container will work....
> 
> Is there an alternative that I don't know about?

Yes, you can do something like this:

```
jobs:
  - name: Big Foo
    plan:
      - aggregate:
        - get: foo_git
          trigger: true

      - task: Build Test
        privileged: true
        params:
          IMAGE_NAME: foo_image
        config:
          platform: linux
          image_resource:
            type: docker-image
            source:
              repository: foo:443/foo/alpine_dind
          inputs:
            - name: foo_git
          outputs:
            - name: docker-images
          run:
            path: /bin/bash
            args:
              - -exc
              - |
                source /docker-lib.sh
                start_docker # Starts Docker in Swarm Mode
                cd foo_git
                make build
                make start # starts a Docker Stack with the build container, and all others needed
                make test # Runs some tests
                docker save foo:latest > ../docker-images/foo.latest.tar

      - put: foo
        params:
          load_repository: foo
          load_file: docker-images/foo.latest.tar
          tag_as_latest: true
        get_params:
          skip_download: true
```
That starts an Alpine container, turns on DinD, builds an image, starts it, runs some tests, saves the image to a file .... and then if everything is OK puts the image (from file) to the Docker repo.

Not every part is there, but the rest of the answers are not too far away.

**Comment by AnianZ** [comment]
This workflow seems to have stopped working with the release of concourse 5.0.0. I can't get the docker-compose in a container setup working any more. It fails with `mount: can't find /sys/fs/cgroup in /proc/mounts` 

Related post in the forum by another person affected: https://discuss.concourse-ci.org/t/docker-compose-inside-concourse-5-no-longer-working/1232

**Comment by meAmidos** [comment]
@AnianZ This might help with the issue: https://github.com/meAmidos/dcind/pull/12

**Comment by AnianZ** [comment]
@meAmidos Thank you! That was exactly what I was looking for. Apparently I was watching the wrong version of dcind.

**Comment by willejs** [comment]
@vito is there any future plan or ideas as to how to tackle this?

**Comment by vito** [comment]
@willejs The situation hasn't really changed from last time. We have too much on our plate and no one from the community has really tried to tackle this and come up with RFCs/proposals, so until that happens this likely won't go anywhere beyond the current solutions. :slightly_frowning_face:

**Comment by Sispheor** [comment]
What about [something like proposed on Github action](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices)?

```yaml
services:
  nginx:
    image: nginx
    # Map port 8080 on the Docker host to port 80 on the nginx container
    ports:
      - 8080:80
  redis:
    image: redis
    # Map TCP port 6379 on Docker host to a random free port on the Redis container
    ports:
      - 6379/tcp
```

The only thing missing in Action is the ability to use a private registry for the image. But otherwise the implementation seems correct.

**Comment by lucasmdrs** [comment]
Most CIs use a similar definition to the on above as shown by @Sispheor , TravisCI, CircleCI, Github Actions, GitlabCI... it doesn't seem like a big feature request.

This makes it harder to run integration tests, that currently only available through some sort of workaround with docker-compose or scripts.

**Comment by jchesterpivotal** [comment]
>  it doesn't seem like a big feature request.

It doesn't _appear_ to be a big request, no.

But I think the history of this issue shows that it's a non-trivial change in the current architecture of Concourse.

It might become easier in future as various architectural refactoring work takes place in order to support Kubernetes as a runtime.

**Comment by lucasmdrs** [comment]
> But I think the history of this issue shows that it's a non-trivial change in the current architecture of Concourse.

By no means I meant to say it's a trivial change, I apologize if I misspelled and seemed that way.

But the history of this issue, the date of opening and the large adoption of this feature in other CI tools, shows that it shouldn't be diminished to a *Core side-road* ("Small features or bug fixes that aren't part of any epic on the roadmap") even less to a *Backburner* priority lane.

I'm here to support the request and keep this running, because it's something I would love to see being payed more attention to, as it's something that prevents me to fully adopt the tool.

**Comment by vito** [comment]
> But the history of this issue, the date of opening and the large adoption of this feature in other CI tools, shows that it shouldn't be diminished to a _Core side-road_ ("Small features or bug fixes that aren't part of any epic on the roadmap") even less to a _Backburner_ priority lane.

I fully agree with this, and it would be great if someone from the community could help out. As I said before, we have too much on our plate right now to take this on ourselves. :/

While this is clearly a highly-sought-after feature, there are other highly-sought-after features which are more important right now because the pain felt without them is much worse: namely, support for branches/pull-requests (see the [v10 roadmap](https://blog.concourse-ci.org/core-roadmap-towards-v10/)) and support for runtimes like Kubernetes. Compared to those feature gaps, this issue is a quality-of-life improvement which has existing workarounds and patterns (though they may not be perfect), which means it has lower priority.

Concourse is a small team and we have to choose our battles. Unfortunately with as many users as we have it's not a matter of simply saying things are high priority, it's a matter of choosing who to disappoint, and hoping the community can pick things up where we can't. :slightly_frowning_face: 

Once we get past those big features we can start to take a deeper look at this. But y'all don't have to wait. Progress in the form of technical proposals and proof-of-concepts would really help get the ball rolling. If you want to support this request, that's really the way.

**Comment by rucciva** [comment]
Hi all, i'm new to concourse, and i have a question, does concourse has some kind of detachable/background task? Regarding this issue, i'm thinking of running something like `docker-compose run --no-deps --use-aliases <service-name>` as concourse (detachable) task for each of the dependency services before running the final test service. As for the docker daemon, i'm thinking of mapping /var/run/docker.sock to tcp port using something like mutagen so that concourse worker can have access to it.

**Comment by danihodovic** [comment]
A Concourse CI resource that executes docker-compose against a remote host:

https://github.com/troykinsella/concourse-docker-compose-resource

**Comment by aoldershaw** [comment]
I put together an [RFC](https://github.com/concourse/rfcs/pull/84) with a proposal for a new `services:` field for `tasks` that looks like:

```yaml
task: integration-tests
file: ci/tasks/test.yml
params:
  POSTGRES_HOST: ((.svc:postgres.host))
  POSTGRES_PORT: ((.svc:postgres.port))
services:
- name: postgres
  file: ci/services/postgres.yml
```

Would love to hear anyone's thoughts! concourse/rfcs#84
